web/source/text/prediction/modelManager.ts

```diff
@@ -63,16 +82,30 @@ namespace com.keyman.text.prediction {
     }
 
     onKeyboardChange(kbdInfo: KeyboardChangeData) {
+      let keyman = com.keyman.singleton;
       let lgCode = kbdInfo['languageCode'];
-
       let model = this.languageModelMap[lgCode];
 
       if(this.currentModel !== model) {
-        this.unloadModel();
+        let stateChange = 0; // Base value; will not remain the same 
+
+        if(this.currentModel) {
+          this.unloadModel();
+          stateChange += 2;
+        }
 
         if(model) {
           this.loadModel(model);
+          stateChange += 1;
         }
+
+        if(stateChange == 0) {
+          console.warn("Unexpected lack of model state change in ModelManager.onKeyboardChange!");
+          return;
+        }
+
+        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : 'changed');
```

Comment:
I found it a little confusing why you declared a `ModelChangeEnum` but then tracking the state with arithmetic. Would this alternative do the same (from ll 90-107)?

```suggestion
      if(this.currentModel !== model) {
        let stateChange: ModelChangeEnum = 'changed';
        
        if(this.currentModel) {
          this.unloadModel();
          stateChange = 'unloaded';
        }
        
        if(model) {
          this.loadModel(model);
          stateChange = 'loaded';
        }
        
        if(stateChange == 'changed') {
          console.warn("...");
          return;
        }
        
       keyman.util.callEvent(ModelManager.EVENT_PREFIX + 'modelchange', stateChange);        
```

Reply:
This won't do the same; `'changed'` is for when _both_ unloading _and_ loading occur.  In your suggestion, it's for when _neither_ occur.

The fact that both can happen together is why I used arithmetic to calculate it internally, as it's easier to work with bit-flags for that.  Of course, those are less understandable outside the method, hence swapping to `string` values for the actual event parameter.

Reply:
So, maybe the following?

```
        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : /* == 3 */ 'changed');
```

Reply:
Perhaps instead:

```suggestion
        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : /* == 3 */ 'changed');
```

Reply:
for clarity, maybe rename `ModelChangeEnum` to 
```
export type ModelStateEnum = 'invalid' | 'loaded' | 'unloaded' | 'changed';
```
and adjust the "state machine"?

Reply:
I feel like that'd indicate that `invalid` is a state that should be handled by the event's callbacks - something that isn't the case.  Is there a reason for wanting to handle `invalid` that I'm missing?

Reply:
To be clear, `'invalid'` _should_ be logically impossible - the `console.warn` is there to facilitate debugging if/when it breaks during development work.  I would consider it a bug if `'invalid'` **ever** occurred in production.